app/src/main/java/edu/nextstep/camp/calculator/MainPresenter.kt

```diff
@@ -1,34 +1,55 @@
 package edu.nextstep.camp.calculator
 
+import android.view.View
+import android.widget.TextView
+import androidx.core.view.isVisible
+import androidx.recyclerview.widget.RecyclerView
 import edu.nextstep.camp.calculator.domain.Calculator
 import edu.nextstep.camp.calculator.domain.Expression
 import edu.nextstep.camp.calculator.domain.Operator
+import edu.nextstep.camp.calculator.domain.Memory
 
-class MainPresenter(view: MainContract.View) : MainContract.Presenter {
-    override var expression: Expression = Expression()
-    override var calculator: Calculator = Calculator()
-    override lateinit var operator: Operator
-    var view = view
+class MainPresenter(private val view: MainContract.View) : MainContract.Presenter {
+    private var expression: Expression = Expression()
+    private var history: Memory = Memory()
+    val calculator: Calculator = Calculator()
 
-    override fun clickOperand(operand: Int) {
-        expression = expression.plus(operand)
+    override fun addToExpression(operand: Int) {
+        expression += operand
         view.showExpression(expression.toString())
     }
 
-    override fun clickOperator(operator: Operator) {
-        expression = expression.plus(operator)
+    override fun addToExpression(operator: Operator) {
+        expression += operator
         view.showExpression(expression.toString())
     }
 
-    override fun clickEqual() {
-        var result = calculator.calculate(expression.toString())
-            ?: return view.showToast("완성되지 않은 수식입니다.")
+    override fun calculateExpression() {
+        val result = calculator.calculate(expression.toString())
+            ?: return view.showError("완성되지 않은 수식입니다.")
+        history += Memory.MemoryItem(expression = expression.toString(), result = result.toString())
         expression = Expression.EMPTY + result
         view.showExpression(expression.toString())
     }
 
-    override fun clickDelete() {
+    override fun removeLastFromExpression() {
         expression = expression.removeLast()
         view.showExpression(expression.toString())
     }
+
+    override fun toggleMemoryView(recyclerView: RecyclerView, textView:TextView) : Boolean {
+        if (recyclerView.isVisible) {
+            recyclerView.visibility = View.INVISIBLE
+            textView.visibility = View.VISIBLE
+            return false
+        }
+        textView.visibility = View.INVISIBLE
+        recyclerView.visibility = View.VISIBLE
+        return true
+    }
```

Comment:
Presenter는 뷰와 협력할때 Contract's View를 통해서만 요청하는게 MVP 아키텍쳐 기본 컨셉입니다.
View 에게 직접 의존성이 있으면 되지 않고 View 작업을 진행하지 않습니다.
현재 toogleMemoryView는 View의 객체를 넘겨 받아 뷰관련 작업을 하고 있는데요. View의 의존성이 없게 변경해보시기 바랍니다.

```suggestion
    override fun toggleMemoryView(isVisibleMemoryView: Boolean) : Boolean {
        when (isVisibleMemoryView) {
            true -> hideMemoryView()
            false -> showMemoryView()
        }
    }
```

Reply:
단순 예시로 전달드린 소스코드입니다. 어떻게 하면 뷰의 의존성을 없애고 표현할 수 있을지 고민해 보시기 바랍니다. 🙏

Reply:
presenter에서 아예 toggle~함수가 필요 없을 것 같습니다!